Generic Presence Detection

rkoshak opened the pull request:
Initial implementation of the Generic Presence Detection example from the forum. I've expanded it somewhat so that it can track both overall presence as well as presence of individuals. I use Scott's automatically generated Rule Triggers trick to make that happen. Configuration takes place through three variables in configuration.py, two lists (list of sensor Items and corresponding proxy Items) as well as the flapping time. 

As I type this message it occurs to me that I could use a single List of tuples which would let me have a different flapping timeout value for each person. I'll await the initial review though before I add that.

CrazyIvan359 commented on Community/GenericPresence/automation/jsr223/python/community/genericpresence/presence.py:
```diff
@@ -0,0 +1,159 @@
+
+        # Item and sensor are the same, cancel the flapping timer if there is
+        # one.
+        global presence_timers # For the love of Pete why?! I get "use before assigned" errors without the global
```
Because you are assigning the value at line 136. It looks like a mistake, you should be doing `presence_timers[proxy_name]` I think

rkoshak replied:
Ah, yes that would explain it. You are right, that is totally a bug. Wonder why this worked when I tested it? I bet I only tested it with one Timer running at a time. I'll test it some more this evening and check in the fix.

rkoshak replied:
OK, this bug should be fixed in the most recent commit. I was able to test live data with multiple Timers yesterday and it work as expected.

CrazyIvan359 commented on Community/GenericPresence/automation/jsr223/python/community/genericpresence/presence.py:
```diff
@@ -0,0 +1,159 @@
+    try:
+        sensor_name = event.itemName
+        proxy_name = presence_config[sensor_name][0]
+        flap_time = presence_config[sensor_name][1]
```
This would be a safer way to do this. You want to catch errors in the above 3 lines only, but you might catch one in the rest of it.

```python
    try:
        sensor_name = event.itemName
        proxy_name = presence_config[sensor_name][0]
        flap_time = presence_config[sensor_name][1]
    except ValueError:
        presence.log.error("{} is not in presence_sensors!"
                           .format(event.itemName))
    except IndexError:
        presence.log.error("presence_items does not have the same number of "
                           "elements as presence_sensors, check your "
                           "configuration!")
    else:
        if items[sensor_name] == items[proxy_name]:
            if (proxy_name in presence_timers.keys()
                    and not presence_timers[proxy_name].hasTerminated()):
                presence.log.debug("{} came home, cancelling the flapping "
                                   "timer.".format(proxy_name))
                presence_timers[proxy_name].cancel()
                del presence_timers[proxy_name]
            return

        # Item and sensor are different, set a flapping timer if the sensor went
        # to OFF.
        if event.itemState == OFF:
            presence.log.debug("Setting the flapping timer for {}"
                               .format(proxy_name))
            presence_timers = ScriptExecution.createTimer(
                DateTime.now().plusMinutes(presence_flap_time),
                lambda: away(presence.log, events, proxy_name, presence_timers))
        elif event.itemState == ON:
            presence.log.debug("{} came home.".format(proxy_name))
            events.sendCommand(proxy_name, "ON")
```

rkoshak replied:
I actually wrote it like that to begin with but I kept getting syntax and runtime errors on the else:. I was supposed to come back and figure out why and I guess I forgot. It seems to be working correctly now. This probably explains why sometimes I would get the error that presence_timers doesn't have a getKeys() method in the unload function. 

This is taking some getting used to. With Rules DSL I know that if VSCode shows no errors the Rule will at least run. With Python I have to run it just to discover some common typos. This is going to be hard a frustrating for users migrating from Rules DSL.

CrazyIvan359 replied:
The problem is the environment. We can't fake the OH script extension context to test the rules in without tremendous effort. At least rules are loaded when the files are modified so you don't have to restart OH every time. Installing pylint can help with obvious stuff, but is also unaware of the OH context

rkoshak commented:
I was experiencing some strange behavior (i.e. Timers not being cancelled when they are supposed to be) so I simplified some of the code and and added a new Rule to synchronize the proxy Items with the current/restored sensor readings at Rule startup. I also added code where the user can set a name metadata on the proxy Item and get more human friendly logs and/or use it in TTS announcements and the like.

volfan6415 commented:
Rather than configured with a list of items could the script be configured to use a group that is defined?

rkoshak commented:
It already works that way. The list is expected to be a list of Groups. In my current setup and in the example in the docstrings, there are three Groups. One generic group that represents when someone is home (gPresent), one specific group that represents when Rich is home (gRichPresent) and one group that  represents when Jenn (gJennPresent) is home. Each of those Groups have more than one member and the script will track each of them separately.

If you only care whether someone is home, you just need the one Group (gPresent). The docs are to show how you can also separately track individual people, not just over all presence.

boehan commented on Community/GenericPresence/automation/jsr223/python/community/genericpresence/presence.py:
```diff
@@ -0,0 +1,159 @@
+    """
+    if not presence_timers:
+        for key in presence_timers.keys():
+            if not presence_timers[key].hasTerminated()
```
```suggestion
            if not presence_timers[key].hasTerminated():
```

rkoshak replied:
This one too was fixed in the commit I forgot to push. I actually changed to use list comprehension instead of nesting the if inside the for loop.

rkoshak commented on Community/GenericPresence/automation/jsr223/python/community/genericpresence/presence.py:
```diff
@@ -0,0 +1,159 @@
+            presence.log.debug("Setting the flapping timer for {}"
+                               .format(proxy_name))
+            presence_timers[proxy_name] = ScriptExecution.createTimer(
+                DateTime.now().plusMinutes(presence_flap_time),
```
Yes it should, and I thought I had already changed that. Turns out I did, I just forgot to push the changes to the PR. Thanks for bringing this to me attention!

rkoshak commented on Community/GenericPresence/automation/jsr223/python/community/genericpresence/presence.py:
```diff
@@ -0,0 +1,159 @@
+"""
+from core.rules import rule
+from core.triggers import when
+from configuration import presence_config
```
That is a known gotcha with how modules work with scripted automation. Changes are not picked up from modules except at OH start time or if you add this reload to the imports. None of the rest of the community libraries appear to include the reload so I didn't either. I certainly don't mind adding it but will wait for Scott to chime in. On an RPi, forcing a recompile of all the modules every time the script loads may not be desirable.

This is talked about a little in the docs: https://openhab-scripters.github.io/openhab-helper-libraries/Python/Reference.html#modifying-and-reloading-packages-or-modules

boehan commented:
@rkoshak is the most recent version in this PR actually working for you? For me there seems to be anything wrong with the lambda function, which I can't get fixed. I always get the following errors when the rule is triggered:
````
2019-11-09 16:02:29.024 [ERROR] [org.quartz.core.JobRunShell         ] - Job DEFAULT.Timer 3 2019-11-09T16:02:29.011+01:00: <function <lambda> at 0x5> threw an unhandled Exception: 
org.python.core.PyException: null
	at org.python.core.Py.NameError(Py.java:290) ~[?:?]
	at org.python.core.PyFrame.getglobal(PyFrame.java:265) ~[?:?]
	at org.python.pycode._pyx33.get_name$1(<script>:78) ~[?:?]
	at org.python.pycode._pyx33.call_function(<script>) ~[?:?]
	at org.python.core.PyTableCode.call(PyTableCode.java:171) ~[?:?]
	at org.python.core.PyBaseCode.call(PyBaseCode.java:139) ~[?:?]
	at org.python.core.PyFunction.__call__(PyFunction.java:413) ~[?:?]
	at org.python.pycode._pyx33.away$6(<script>:127) ~[?:?]
	at org.python.pycode._pyx33.call_function(<script>) ~[?:?]
	at org.python.core.PyTableCode.call(PyTableCode.java:171) ~[?:?]
	at org.python.core.PyBaseCode.call(PyBaseCode.java:308) ~[?:?]
	at org.python.core.PyFunction.function___call__(PyFunction.java:471) ~[?:?]
	at org.python.core.PyFunction.__call__(PyFunction.java:466) ~[?:?]
	at org.python.core.PyFunction.__call__(PyFunction.java:456) ~[?:?]
	at org.python.pycode._pyx33.f$5(<script>:106) ~[?:?]
	at org.python.pycode._pyx33.call_function(<script>) ~[?:?]
	at org.python.core.PyTableCode.call(PyTableCode.java:171) ~[?:?]
	at org.python.core.PyBaseCode.call(PyBaseCode.java:125) ~[?:?]
	at org.python.core.PyFunction.__call__(PyFunction.java:403) ~[?:?]
	at org.python.core.PyFunction.__call__(PyFunction.java:398) ~[?:?]
	at org.python.core.PyFunction.invoke(PyFunction.java:533) ~[?:?]
	at com.sun.proxy.$Proxy367.apply(Unknown Source) ~[?:?]
	at org.eclipse.smarthome.model.script.internal.actions.TimerExecutionJob.execute(TimerExecutionJob.java:48) ~[?:?]
	at org.quartz.core.JobRunShell.run(JobRunShell.java:202) [182:org.openhab.core.scheduler:2.5.0.M4]
	at org.quartz.simpl.SimpleThreadPool$WorkerThread.run(SimpleThreadPool.java:573) [182:org.openhab.core.scheduler:2.5.0.M4]
2019-11-09 16:02:29.078 [ERROR] [org.quartz.core.ErrorLogger         ] - Job (DEFAULT.Timer 3 2019-11-09T16:02:29.011+01:00: <function <lambda> at 0x5> threw an exception.
org.quartz.SchedulerException: Job threw an unhandled exception.
	at org.quartz.core.JobRunShell.run(JobRunShell.java:213) [182:org.openhab.core.scheduler:2.5.0.M4]
	at org.quartz.simpl.SimpleThreadPool$WorkerThread.run(SimpleThreadPool.java:573) [182:org.openhab.core.scheduler:2.5.0.M4]
Caused by: org.python.core.PyException
	at org.python.core.Py.NameError(Py.java:290) ~[?:?]
	at org.python.core.PyFrame.getglobal(PyFrame.java:265) ~[?:?]
	at org.python.pycode._pyx33.get_name$1(<script>:78) ~[?:?]
	at org.python.pycode._pyx33.call_function(<script>) ~[?:?]
	at org.python.core.PyTableCode.call(PyTableCode.java:171) ~[?:?]
	at org.python.core.PyBaseCode.call(PyBaseCode.java:139) ~[?:?]
	at org.python.core.PyFunction.__call__(PyFunction.java:413) ~[?:?]
	at org.python.pycode._pyx33.away$6(<script>:127) ~[?:?]
	at org.python.pycode._pyx33.call_function(<script>) ~[?:?]
	at org.python.core.PyTableCode.call(PyTableCode.java:171) ~[?:?]
	at org.python.core.PyBaseCode.call(PyBaseCode.java:308) ~[?:?]
	at org.python.core.PyFunction.function___call__(PyFunction.java:471) ~[?:?]
	at org.python.core.PyFunction.__call__(PyFunction.java:466) ~[?:?]
	at org.python.core.PyFunction.__call__(PyFunction.java:456) ~[?:?]
	at org.python.pycode._pyx33.f$5(<script>:106) ~[?:?]
	at org.python.pycode._pyx33.call_function(<script>) ~[?:?]
	at org.python.core.PyTableCode.call(PyTableCode.java:171) ~[?:?]
	at org.python.core.PyBaseCode.call(PyBaseCode.java:125) ~[?:?]
	at org.python.core.PyFunction.__call__(PyFunction.java:403) ~[?:?]
	at org.python.core.PyFunction.__call__(PyFunction.java:398) ~[?:?]
	at org.python.core.PyFunction.invoke(PyFunction.java:533) ~[?:?]
	at com.sun.proxy.$Proxy367.apply(Unknown Source) ~[?:?]
	at org.eclipse.smarthome.model.script.internal.actions.TimerExecutionJob.execute(TimerExecutionJob.java:48) ~[?:?]
	at org.quartz.core.JobRunShell.run(JobRunShell.java:202) ~[?:?]
	... 1 more
````
any ideas?

rkoshak commented:
I've been running with it since before I posted it here. The lambda itself is as simple as can be. I'm not sure what could be throwing an exception from there. Put everything in away() inside a try/except and log out the contents of the exception.

```
def away(log, events, proxy_name, item_name, flap_time):
    """
    Called when a person has been detected away for more than the flapping time.
    Log out the away state and command the Item to OFF.
    Arguments:
        - log: Logger passed in from the presence Rule.
        - events: Access to the events Object so we can call sendCommand.
        - proxy_name: Name of the proxy Item that we want to set to away.
        - item_name: Name of the sensor Item that detects presence.
        - flap_time: Number of minutes we wait before marking the person away.
    """
    try:
        if items[proxy_name] == OFF:
            log.warn("Presence away timer expired but {} is already OFF!"
                     .format(proxy_name))
        elif items[item_name] == ON:
            log.warn("Presence away timer expired but {} is ON!".format(item_name))
        else:
            log.info("{} has been away for {} minutes, setting to away."
                    .format(get_name(proxy_name),flap_time))
            events.sendCommand(proxy_name, "OFF")
    except:
        import sys
        log.error("Error occurred in away: {}".format(sys.exec_info()[0]))
```

That should tell us what the error is and on which line it occurs. I can't imagine what it would be though. Perhaps `proxy_name` is None instead of the name of the proxy Item in which case there may be something wrong with your configuration.py settings.

rkoshak commented:
This PR is ready for review

boehan commented:
> I've been running with it since before I posted it here. The lambda itself is as simple as can be. I'm not sure what could be throwing an exception from there. Put everything in away() inside a try/except and log out the contents of the exception.
> 
> ```
> def away(log, events, proxy_name, item_name, flap_time):
>     """
>     Called when a person has been detected away for more than the flapping time.
>     Log out the away state and command the Item to OFF.
>     Arguments:
>         - log: Logger passed in from the presence Rule.
>         - events: Access to the events Object so we can call sendCommand.
>         - proxy_name: Name of the proxy Item that we want to set to away.
>         - item_name: Name of the sensor Item that detects presence.
>         - flap_time: Number of minutes we wait before marking the person away.
>     """
>     try:
>         if items[proxy_name] == OFF:
>             log.warn("Presence away timer expired but {} is already OFF!"
>                      .format(proxy_name))
>         elif items[item_name] == ON:
>             log.warn("Presence away timer expired but {} is ON!".format(item_name))
>         else:
>             log.info("{} has been away for {} minutes, setting to away."
>                     .format(get_name(proxy_name),flap_time))
>             events.sendCommand(proxy_name, "OFF")
>     except:
>         import sys
>         log.error("Error occurred in away: {}".format(sys.exec_info()[0]))
> ```
> 
> That should tell us what the error is and on which line it occurs. I can't imagine what it would be though. Perhaps `proxy_name` is None instead of the name of the proxy Item in which case there may be something wrong with your configuration.py settings.

That did not help, but I think I found the problem. After adding `from core.metadata import get_value` to the imports it seems to be working now (will have to test this over the next days).

rkoshak commented:
I'm not certain why that import is missing in the first place, but that would cause problems.  
When I look at my copy it has the import but the import is clearly missing here. I messed up somewhere.


Unfortunately I can't fix it until this PR has been reviewed.

rkoshak commented:
I've basically completely rewritten this so am closing this PR. When reviewing submissions to the helper library becomes a priority I'll resubmit. In the mean time I'll keep in in my own repo where I can continue development.
